frontend/test/metabase/visualizations/lib/settings/visualization.unit.spec.js

```diff
@@ -58,6 +58,53 @@ describe("visualization_settings", () => {
           }),
         ));
     });
+    describe("table.columns", () => {
+      it("should populate an empty table.columns based on data columns", () => {
+        const cols = ["foo", "bar", "baz"].map(name => NumberColumn({ name }));
+        const card = { display: "table", visualization_settings: {} };
+        const settings = getComputedSettingsForSeries([
+          { card, data: { cols } },
+        ]);
+        expect(settings["table.columns"].map(c => c.name)).toEqual([
+          "foo",
+          "bar",
+          "baz",
+        ]);
+      });
+      it("should add to table.columns based on new data columns", () => {
+        const cols = ["foo", "bar", "baz"].map(name => NumberColumn({ name }));
+        const card = {
+          display: "table",
+          visualization_settings: {
+            "table.columns": [{ name: "foo" }, { name: "bar" }],
+          },
+        };
+        const settings = getComputedSettingsForSeries([
+          { card, data: { cols } },
+        ]);
+        expect(settings["table.columns"].map(c => c.name)).toEqual([
+          "foo",
+          "bar",
+          "baz",
+        ]);
+      });
+      it("should remove from table.columns if not present in the data's columns", () => {
```

Comment:
Do we want this behavior? If someone deletes a column from a native query should we drop it from the settings?

Reply:
@paulrosenzweig Isn't that similar to the chart settings, which you talked about in issue 10226?

Reply:
Yeah good point! Here I'm not dropping the entire setting when it's invalid. Instead, I just tweaked the definition of "invalid" so that it would update to something that is valid. So, I suppose with 10226 we'd need a plan for when something can be regenerated to be valid vs when it should be dropped entirely.